src/condition/ssrf_detector.cpp

```diff
@@ -0,0 +1,222 @@
+// Unless explicitly stated otherwise all files in this repository are
+// dual-licensed under the Apache-2.0 License or BSD-3-Clause License.
+//
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2021 Datadog, Inc.
+
+#include "condition/ssrf_detector.hpp"
+#include "exception.hpp"
+#include "iterator.hpp"
+#include "uri_utils.hpp"
+#include "utils.hpp"
+#include <iostream>
+
+using namespace std::literals;
+
+namespace ddwaf {
+
+namespace {
+
+constexpr std::array<std::string_view, 10> dangerous_ips{"169.254.0.0/16", "127.0.0.1/32",
+    "10.0.0.0/8", "172.16.0.0/12", "192.168.0.0/16", "100.64.0.0/10", "::1/128", "fc00::/7",
+    "fe80::/10", "2001:db8:1234:1a00::/56"};
+
+constexpr const std::array<std::string_view, 10> dangerous_domains{
+    "metadata.google", "burpcollaborator.net", ".local", ".internal", "ram.aliyuncs.com",
+    "ifconfig.pro", "localhost", "fuf.me", "localtest.me", "ulh.us" // Legacy domains
+};
+
+constexpr const std::array<std::string_view, 4> authorised_schemes{"https", "http", "ftps", "ftp"};
+
+constexpr const auto &npos = std::string_view::npos;
+
+using ssrf_result = std::optional<std::pair<std::string, std::vector<std::string>>>;
+
+bool detect_parameter_injection(
```

Comment:
I can see many calls to `find` which means param will be iterated over and over. I wonder if it would be better to do a controlled iteration of the string finding they key characters rather the iterating the same many times

Reply:
Perhaps although if the logic is hard to follow now, that wouldn't make it any easier...